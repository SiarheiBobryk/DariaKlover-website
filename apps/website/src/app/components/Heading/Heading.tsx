import { ElementType, ReactNode, Ref, forwardRef } from 'react';

import Typography from '@mui/material/Typography';

export interface HeadingProps {
  children?: ReactNode;
  component?: ElementType;
  variant?:
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6'
    | 'subtitle1'
    | 'subtitle2'
    | 'body1'
    | 'body2'
    | 'caption'
    | 'button'
    | 'overline'
    | 'inherit'
    | undefined;
}

const Heading = forwardRef(function Heading(props: HeadingProps, ref: Ref<HTMLParagraphElement>) {
  const { children, component = 'h1', variant = 'h5', ...other } = props;

  return (
    <Typography ref={ref} component={component} variant={variant} gutterBottom {...other}>
      {children}
    </Typography>
  );
});

export default Heading;
