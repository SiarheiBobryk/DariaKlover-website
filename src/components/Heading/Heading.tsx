import * as React from 'react';

import Typography from '@mui/material/Typography';

export interface HeadingProps {
  children?: React.ReactNode;
  component?: React.ElementType;
  variant?:
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6'
    | 'subtitle1'
    | 'subtitle2'
    | 'body1'
    | 'body2'
    | 'caption'
    | 'button'
    | 'overline'
    | 'inherit'
    | undefined;
}

function Heading(props: HeadingProps) {
  const { children, component = 'h2', variant = 'h5', ...other } = props;

  return (
    <Typography component={component} variant={variant} {...other}>
      {children}
    </Typography>
  );
}

export default Heading;
