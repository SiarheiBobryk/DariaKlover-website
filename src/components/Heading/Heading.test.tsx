import * as React from 'react';

import { axe, toHaveNoViolations } from 'jest-axe';
import { render, RenderResult, screen } from '@testing-library/react';

import { typographyClasses } from '@mui/material/Typography';

import Heading from './Heading';

expect.extend(toHaveNoViolations);

describe('<Heading />', () => {
  describe('conformance', () => {
    it('should match a11y standards', async () => {
      const { container }: RenderResult = render(<Heading>Lorem Ipsulum</Heading>);

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('should forward ref', () => {
      const ref = React.createRef();
      const { container } = render(<Heading ref={ref} />);
      expect(container.firstChild).toBe(ref.current);
    });
  });

  describe('Elements inspection', () => {
    it('should be the `h2` HTML element by default', () => {
      render(<Heading>Дарья Кловер</Heading>);

      const h2: HTMLHeadingElement | null = document.querySelector('h2');
      expect(h2).toBeInTheDocument();
    });

    it('should be the `h1` HTML element', () => {
      render(<Heading component="h1">Дарья Кловер</Heading>);

      const h1: HTMLHeadingElement | null = document.querySelector('h1');
      expect(h1).toBeInTheDocument();
    });

    it('should have the CSS classes from `h5` Typography variant by default', () => {
      render(<Heading>Дарья Кловер</Heading>);

      const heading: HTMLHeadingElement = screen.getByText('Дарья Кловер');
      expect(heading).toHaveClass(typographyClasses.h5);
    });

    it('should have the CSS classes from `h3` Typography variant', () => {
      render(<Heading variant="h3">Дарья Кловер</Heading>);

      const heading: HTMLHeadingElement = screen.getByText('Дарья Кловер');
      expect(heading).toHaveClass(typographyClasses.h3);
    });

    it('should have gutter', () => {
      render(<Heading variant="h3">Дарья Кловер</Heading>);

      const heading: HTMLHeadingElement = screen.getByText('Дарья Кловер');
      expect(heading).toHaveStyle({
        marginBottom: '0.35em',
      });
    });
  });
});
