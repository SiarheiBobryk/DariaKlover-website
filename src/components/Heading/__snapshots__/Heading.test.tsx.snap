// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> snapshotting should render correctly with a content 1`] = `
<h2
  className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
>
  Lorem Ipsulum
</h2>
`;

exports[`<Heading /> snapshotting should render correctly without a content 1`] = `
<h2
  className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
/>
`;
